// @microsoft/fast-foundation/dist/esm/tooltip/tooltip.options.js
/**
 * Enumerates possible tooltip positions
 *
 * @public
 */
export var TooltipPosition;
(function (TooltipPosition) {
    /**
     * The tooltip is positioned above the element
     */
    TooltipPosition["top"] = "top";
    /**
     * The tooltip is positioned to the right of the element
     */
    TooltipPosition["right"] = "right";
    /**
     * The tooltip is positioned below the element
     */
    TooltipPosition["bottom"] = "bottom";
    /**
     * The tooltip is positioned to the left of the element
     */
    TooltipPosition["left"] = "left";
    /**
     * The tooltip is positioned before the element
     */
    TooltipPosition["start"] = "start";
    /**
     * The tooltip is positioned after the element
     */
    TooltipPosition["end"] = "end";
    /**
     * The tooltip is positioned above the element and to the left
     */
    TooltipPosition["topLeft"] = "top-left";
    /**
     * The tooltip is positioned above the element and to the right
     */
    TooltipPosition["topRight"] = "top-right";
    /**
     * The tooltip is positioned below the element and to the left
     */
    TooltipPosition["bottomLeft"] = "bottom-left";
    /**
     * The tooltip is positioned below the element and to the right
     */
    TooltipPosition["bottomRight"] = "bottom-right";
    /**
     * The tooltip is positioned above the element and to the left
     */
    TooltipPosition["topStart"] = "top-start";
    /**
     * The tooltip is positioned above the element and to the right
     */
    TooltipPosition["topEnd"] = "top-end";
    /**
     * The tooltip is positioned below the element and to the left
     */
    TooltipPosition["bottomStart"] = "bottom-start";
    /**
     * The tooltip is positioned below the element and to the right
     */
    TooltipPosition["bottomEnd"] = "bottom-end";
})(TooltipPosition || (TooltipPosition = {}));
